fix: Fixes when ESCPlayer is not selectable or interactive manage click through (reopen)

StraToN opened the pull request:
PR #609 was merged a bit too fast/

Supersedes https://github.com/godot-escoria/escoria-demo-game/pull/605
Fixes https://github.com/godot-escoria/escoria-issues/issues/104

This change verifies the content of the hover stack before performing the action.

Adds a new test room 18 that allows testing superposed ESCItems.

If the clicked item is an unselectable ESCPlayer then we consider the action went through it and applies to:

    the item behind (if any in the hover stack)
    on the background if no item is behind

To test:

1. ESCPlayer superposed to ESCItem
    set mark as NOT selectable (uncheck in the inspector)
    run room 1
    hover mark should do nothing
    left/right/double click through mark
    put mark in front of and item on the wall and do the same (left/right/double click on the item through Mark)

2. Superposed ESCItems
    run room18

balloonpopper commented:
For clarity, I'm separating the above issues into 3. 
Issue 1 = mark should be click through when selectable=false
issue 2 = tooltip is not updating after you completely move off, and on the collision shape
issue 3 = pausing and unpausing doesn't restore the tooltip.

Issue 1 : Yes I can select through Mark now without him moving when I right click the painting, however I'm seeing another issue I don't believe should be there. If you right click the painting, the tooltip is visible. Mark says his line of text. The tooltip disappears when Mark finishes speaking. Right clicking will not work as the tooltip is no longer active - I need to move off/on the painting to get the tooltip to reappear. I can now right click again (which works correctly).
I don't believe the tooltip should disappear.

Issue 2 - I start room 18 and put the mouse in a spot where it is over the lowest collision shape. I right click to remove the upper collision shape. Once it's animation completes, I can't get the tooltip to register item 2 until I move off item 3's collision shape. Moving the mouse a small amount does not correct the tooltip, it only reactivates when I move off the area where the lowest shape is.

Issue 3) I can live with issue 3 as long as a small mouse movement restores the tooltip. I'm not sure how I would implement the solution you suggest because you'd need to confirm where the mouse was when unpausing to see if the tooltip was still correct - and I don't know how you'd do that in code. Anyway, if it's UI specific it can be ignored.

balloonpopper commented:
@StraToN What is happening in relation to the above issues? I'm not sure what the decision was sorry.
Additionally, just retested, i'll list this as issue 4 to not confuse any issues.
Issue 4) Start room 1 (f6). Right click on the bull's nose. Right click to make the text go really fast.
Now right and left clicks around the room don't work (except on the chicken picture). Moving the mouse around the room the tooltip continues to say "artwork" until you highlight the player or chicken picture.

StraToN commented:
> @StraToN What is happening in relation to the above issues? I'm not sure what the decision was sorry. 

Sorry for the very late answer.
Issue 1 = mark should be click through when selectable=false: fixed
Issue 2 = tooltip is not updating after you completely move off, and on the collision shape: should be fixed now
Issue 3 = pausing and unpausing doesn't restore the tooltip. : UI specific

> Additionally, just retested, i'll list this as issue 4 to not confuse any issues. 
> Issue 4) Start room 1 (f6). Right click on the bull's nose. Right click to make the text go really fast. Now right and left clicks around the room don't work (except on the chicken picture). Moving the mouse around the room the tooltip continues to say "artwork" until you highlight the player or chicken picture.

I've been trying to replicate you scenario as you described but I didn't manage to get right and left clicks around the room not to work. Maybe we can setup a small meeting so we can try to have me replicating it?

balloonpopper commented:
I still see the same issue @StraToN. Run room 1 using F6. Right click the first picture. Wait for the text to finish (it doesn't matter if you speed up the text or just wait for the slow text to complete). Click on the floor somewhere - he won't walk there.

BHSDuncan commented:
I can confirm that I can reproduce using the instructions above.

StraToN commented:
Thanks! Should be fixed now, please retest.

balloonpopper commented:
I retested. 
Issue 1) In room 18 I can remove the top item, but the middle item won't move.

```
ESC (2022-09-01T22:37:51) D esc_action_manager.gd: Item upper_item is not interactive.
ESC (2022-09-01T22:37:51) D esc_inputs_manager.gd: Ignoring mouse entering item upper_item.
ESC (2022-09-01T22:37:51) I esc_inputs_manager.gd: Item focused: middle_item
ESC (2022-09-01T22:37:51) I esc_inputs_manager.gd: Item focused: bottom_item
ESC (2022-09-01T22:37:52) D esc_action_manager.gd: Item upper_item is not interactive.
ESC (2022-09-01T22:37:52) D esc_inputs_manager.gd: Ignoring right click on upper_item with event [InputEventMouseButton:2932].
```
Issue 2) Restart godot from scratch. Load room 2. Start it with f6. Left click somewhere to walk on the left platform. When he gets there, left click somewhere else, you can't move.
According to debug, the mouse click isn't registered.
If I click on Mark, that frees up the mouse, and I can make him move on the next click, then the mouse locks up again until I click on Mark.

If I turn "selectable" off on Mark, the problem goes away.

StraToN commented:
> Issue 1) In room 18 I can remove the top item, but the middle item won't move.

I was able to reproduce. Issue fixed.

>Issue 2) 

Same: I was able to reproduce and fix. 

I took the opportunity to rebase on develop. We should now be ok, waiting for retest. Thanks a lot @balloonpopper  :+1:

balloonpopper commented:
I can now remove the top two items. Two things I've noticed.
1) If I click to remove the top item, then click somewhere that is where the top item used to be (but not where the middle item is) I get
```
ESC (2022-09-18T21:54:25) D esc_event_manager.gd: Event use ended with return code 0.
ESC (2022-09-18T21:54:28) D esc_action_manager.gd: Item upper_item is not interactive.
ESC (2022-09-18T21:54:28) D esc_inputs_manager.gd: Ignoring right click on upper_item with event [InputEventMouseButton:2913].
ESC (2022-09-18T21:54:28) I esc_inputs_manager.gd: Right click on background at (349, 157).
```
So firstly, should non-interactive items show that you tried to click them? I was expecting it should be totally ignored. 
Would this still swallow the click if you were trying to move your avatar somewhere that is now covered by an inactive item?

2) Every time I click the bottom item (after removing top and middle) I get 3 errors. 
```
ESC (2022-09-18T21:58:00) I esc_item.gd: Current game state doesn't accept interactions.
ESC (2022-09-18T21:58:00) I esc_item.gd: Current game state doesn't accept interactions.
ESC (2022-09-18T21:58:00) I esc_item.gd: Current game state doesn't accept interactions.
```
Should I get any errors? If I should, there should only be 1 instead of 3.

balloonpopper commented:
Issue 3) Tested the above scenario 
![image](https://user-images.githubusercontent.com/5151242/190901169-ab41465c-9f13-4fe7-93ab-fa82861f8954.png)
I copied a painting in room 1, put it on the navigation shape and made it non-interactive. If you right click the new painting, you can't walk where it is.

StraToN commented:
> 1. If I click to remove the top item, then click somewhere that is where the top item used to be (but not where the middle item is) I get
> 
> ```
> ESC (2022-09-18T21:54:25) D esc_event_manager.gd: Event use ended with return code 0.
> ESC (2022-09-18T21:54:28) D esc_action_manager.gd: Item upper_item is not interactive.
> ESC (2022-09-18T21:54:28) D esc_inputs_manager.gd: Ignoring right click on upper_item with event [InputEventMouseButton:2913].
> ESC (2022-09-18T21:54:28) I esc_inputs_manager.gd: Right click on background at (349, 157).
> ```
> 
> So firstly, should non-interactive items show that you tried to click them? I was expecting it should be totally ignored. 

If you didn't expect these messages, I'm fine removing them. They were debug messages, only here to inform gamedev that an item was clicked and ignored because it's interactive.

Therefore, I removed these messages. They'll be easy to re-add if someone needs them.

> Would this still swallow the click if you were trying to move your avatar somewhere that is now covered by an inactive item?

I expect this is not going to happen.



> 2. Every time I click the bottom item (after removing top and middle) I get 3 errors.
> 
> 
> ```
> ESC (2022-09-18T21:58:00) I esc_item.gd: Current game state doesn't accept interactions.
> ESC (2022-09-18T21:58:00) I esc_item.gd: Current game state doesn't accept interactions.
> ESC (2022-09-18T21:58:00) I esc_item.gd: Current game state doesn't accept interactions.
> ```
> 
> Should I get any errors? If I should, there should only be 1 instead of 3.

I couldn't replicate this issue in current state of my branch. I have force-pushed after a rebase on `develop`, maybe that made the error go away?

>  Issue 3) Tested the above scenario
>  I copied a painting in room 1, put it on the navigation shape and made it non-interactive. If you right click the new painting, you can't walk where it is.

Good catch. Fixed.

balloonpopper commented:
Issues found on the current commit
1) If I move my mouse around the middle of the object in room 18, the item in focus (according to the logs) changes. I'm expecting it to stay as the top item.
```
ESC (2022-11-08T23:38:48) I esc_inputs_manager.gd: Item unfocused: upper_item
ESC (2022-11-08T23:38:48) I esc_inputs_manager.gd: Item focused: bottom_item
```

2) An extension of point 1 above. You can see in the below that the focus was the middle item (I had my mouse in a position covering the top and middle item, but not the bottom), then after clicking the middle_item was focused a second time, even though it was already the focus.
```
ESC (2022-11-08T23:52:52) I esc_inputs_manager.gd: Item focused: upper_item
ESC (2022-11-08T23:52:53) I esc_inputs_manager.gd: Item focused: middle_item
ESC (2022-11-08T23:52:54) I esc_inputs_manager.gd: Item upper_item right clicked with event [InputEventMouseButton:3137].
ESC (2022-11-08T23:52:54) I esc_action_manager.gd: item_right_click on item upper_item.
ESC (2022-11-08T23:52:54) I esc_action_manager.gd: upper_item to perform event [InputEventMouseButton:3137].
ESC (2022-11-08T23:52:54) I esc_action_manager.gd: Checking if action 'use' on '[Reference:2052]' is valid...
ESC (2022-11-08T23:52:54) D esc_event_manager.gd: Queueing event use in channel _front.
ESC (2022-11-08T23:52:54) D esc_event_manager.gd: Popping event 'use' from background queue _front to source res://game/rooms/room18/esc/upper_item.esc.
ESC (2022-11-08T23:52:54) D esc_event.gd: Event use started.
ESC (2022-11-08T23:52:54) D esc_command.gd: Running command set_interactive with parameters [upper_item, False].
ESC (2022-11-08T23:52:54) I esc_inputs_manager.gd: Item focused: middle_item
```

3) When moving the mouse right to the middle of the objects (so the mouse is over where all 3 objects are) then back up a little bit (to where the mouse would be over the upper and middle items, but not over the bottom item.
```
ESC (2022-11-08T23:55:42) I esc_inputs_manager.gd: Item focused: upper_item
ESC (2022-11-08T23:55:43) I esc_inputs_manager.gd: Item focused: middle_item
ESC (2022-11-08T23:55:43) I esc_inputs_manager.gd: Item focused: bottom_item
ESC (2022-11-08T23:55:45) I esc_inputs_manager.gd: Item unfocused: upper_item
```

4) I'm still seeing the issue mentioned in the last post above. The first time I right click the bottom item it tries to "say" a line, but if I right click again (without moving the mouse) then every time I right click I get the "Current game state doesn't accept interactions." error
```
ESC (2022-11-08T23:40:25) D esc_event_manager.gd: Popping event 'use' from background queue _front to source res://action_defaults.esc.
ESC (2022-11-08T23:40:25) D esc_event.gd: Event use started.
ESC (2022-11-08T23:40:25) D esc_command.gd: Running command say with parameters [current_player, "I can't use this!", ].
ESC (2022-11-08T23:40:25) W say.gd: [say]: No player item in the current scene was registered and the say command was encountered.
ESC (2022-11-08T23:40:25) D esc_command.gd: [say] Return code: 1.
ESC (2022-11-08T23:40:25) D esc_event_manager.gd: Event 'use' ended with return code 1.
ESC (2022-11-08T23:40:36) I esc_item.gd: Current game state doesn't accept interactions.
ESC (2022-11-08T23:40:36) I esc_item.gd: Current game state doesn't accept interactions.
ESC (2022-11-08T23:40:36) I esc_item.gd: Current game state doesn't accept interactions.
```

5) In room 1 I copied the first painting, changed its global_id = r1_wall_item3, turned interactive=off, and moved it in the room.
I can right click the object which moves the player where you right click, however if I click the floor somewhere else, I can't walk to the new location. (I can fix this by right clicking one of the original paintings or a door, but I can't just walk somewhere else in the room)
![image](https://user-images.githubusercontent.com/5151242/200571353-2f0827d8-1daa-48e9-9dad-4721d64c76ab.png)

StraToN commented on addons/escoria-core/game/core-scripts/esc/types/esc_object.gd:
```diff
@@ -97,6 +97,9 @@ func _get_interactive() -> bool:
 	if "is_interactive" in self.node:
 		self.node.is_interactive = value
+		if not value:
+			(escoria.game_scene.tooltip_node as ESCTooltip).clear()
```
This is dangerous if game scene has no tooltip. See if there is a method to call in ESCGame for this instead

StraToN commented on addons/escoria-core/game/esc_inputs_manager.gd:
```diff
@@ -159,13 +179,70 @@ func clear_stack():
 func _on_left_click_on_bg(position: Vector2) -> void:
-	if input_mode == INPUT_ALL and hotspot_focused.empty():
+	if input_mode == INPUT_ALL: # and hotspot_focused.empty():
+		hotspot_focused = ""
```
Possible to remove this `hotspot_focused` attribute since we now have `_hovered_element`?

balloonpopper commented:
@StraToN : Issues 1,2,3 and 5 from the above are pretty much fixed. 4 is still there. What I'm seeing now :

1) In room 18, when I mouse over the objects, only the top_item focuses which is good, but after you remove the top item the bottom item comes into focus before the middle item.
```
ESC (2022-11-15T23:18:36) D esc_object.gd: State "folded" set. Matching non-immediate animation executing.
ESC (2022-11-15T23:18:36) D esc_command.gd: [set_state] Return code: 0.
ESC (2022-11-15T23:18:36) D esc_event_manager.gd: Event 'use' ended with return code 0.
ESC (2022-11-15T23:18:36) I esc_inputs_manager.gd: Item focused: bottom_item
ESC (2022-11-15T23:18:36) I esc_inputs_manager.gd: Item focused: middle_item
```

2) After I've removed the top and middle items, you can see if I move my mouse away that the bottom_item is registering correctly
```
ESC (2022-11-15T23:20:24) I esc_inputs_manager.gd: Item focused: bottom_item
ESC (2022-11-15T23:20:27) I esc_inputs_manager.gd: Item unfocused: bottom_item
```
If I now right click the bottom_item, you'll see that it has an error with the say command, then when I move my mouse away it no longer knows the focused item (see last line of the output)
```
ESC (2022-11-15T23:20:29) I esc_action_manager.gd: item_right_click on item bottom_item.
ESC (2022-11-15T23:20:29) I esc_action_manager.gd: bottom_item to perform event [InputEventMouseButton:4069].
ESC (2022-11-15T23:20:29) I esc_action_manager.gd: Checking if action 'use' on '[Reference:1996]' is valid...
ESC (2022-11-15T23:20:29) D esc_event_manager.gd: Queueing event use in channel _front.
ESC (2022-11-15T23:20:29) D esc_event_manager.gd: Popping event 'use' from background queue _front to source res://action_defaults.esc.
ESC (2022-11-15T23:20:29) D esc_event.gd: Event use started.
ESC (2022-11-15T23:20:29) D esc_command.gd: Running command say with parameters [current_player, "I can't use this!", ].
ESC (2022-11-15T23:20:29) W say.gd: [say]: No player item in the current scene was registered and the say command was encountered.
ESC (2022-11-15T23:20:29) D esc_command.gd: [say] Return code: 1.
ESC (2022-11-15T23:20:29) D esc_event_manager.gd: Event 'use' ended with return code 1.
ESC (2022-11-15T23:20:33) I esc_inputs_manager.gd: Item unfocused:
```
If I then mouse over it again, it knows the item is the bottom item again
```
ESC (2022-11-15T23:22:36) I esc_inputs_manager.gd: Item focused: bottom_item
ESC (2022-11-15T23:22:37) I esc_inputs_manager.gd: Item unfocused: bottom_item
```
So there's a problem with it losing the focus temporarily

3) If I now right click the bottom_item again I get the 3 errors seen before. Every right click gives 3 errors.
```
ESC (2022-11-15T23:24:01) I esc_inputs_manager.gd: Item focused: bottom_item
ESC (2022-11-15T23:24:03) I esc_item.gd: Current game state doesn't accept interactions.
ESC (2022-11-15T23:24:03) I esc_item.gd: Current game state doesn't accept interactions.
ESC (2022-11-15T23:24:03) I esc_item.gd: Current game state doesn't accept interactions.
```

Room 19 functionality and messages all look good.

StraToN commented:
@balloonpopper 
I was able to reproduce point 1 only once, it doesn't seem to happen every time. Can you confirm this? Do you have specific conditions to reproduce (apart from your description)

Other points (2 and 3) shoud be resolved by [20ce8f2](https://github.com/godot-escoria/escoria-demo-game/pull/613/commits/20ce8f2415356cb0d917b043b795a696003938ce)

balloonpopper commented:
1) For issue 1, I've pushed a commit to this branch. If you right click in the exact centre of the box I added to room 18, the issue 1 above always happens. It seems very picky about where on the object it happens

2) Fixed

3) Fixed

4) New issue - not sure that this will be fixable :
If you alt tab away from the running game and back in (with the mouse over the middle of the object(s)), you get multiple focussed items in the debug. It doesn't break anything that I can tell though.

ESC (2022-11-18T17:17:13) I esc_inputs_manager.gd: Item focused: upper_item
<Tab away and back>
ESC (2022-11-18T17:17:14) I esc_inputs_manager.gd: Item unfocused: upper_item
ESC (2022-11-18T17:17:16) I esc_inputs_manager.gd: Item focused: middle_item
ESC (2022-11-18T17:17:16) I esc_inputs_manager.gd: Item focused: upper_item
<Tab away and back>
ESC (2022-11-18T17:17:17) I esc_inputs_manager.gd: Item unfocused: upper_item
ESC (2022-11-18T17:17:18) I esc_inputs_manager.gd: Item focused: bottom_item
ESC (2022-11-18T17:17:18) I esc_inputs_manager.gd: Item focused: middle_item
ESC (2022-11-18T17:17:18) I esc_inputs_manager.gd: Item focused: upper_item
ESC (2022-11-18T17:17:20) I esc_inputs_manager.gd: Item unfocused: upper_item

StraToN commented:
@balloonpopper 
I managed to reproduce this: it only happens when the cursor passes from the verbs panel at the bottom to the items in the room. If the cursor enters the items from the outside of the items, but in the room, then the issue doesn't occur.
This is actually the same issue as 4) : mouse_entered signal happen from bottom to top :/ 
I am really not certain what I can do about this.

balloonpopper commented:
It doesn't appear to break anything anyway, so happy to approve it.

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc/esc_action_manager.gd:
```diff
@@ -121,12 +121,11 @@ func do(action: int, params: Array = [], can_interrupt: bool = false) -> void:
 						"item_left_click on item %s." % params[0]
 					)
-
+					
```
Remove whitespace

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc/esc_action_manager.gd:
```diff
@@ -121,12 +121,11 @@ func do(action: int, params: Array = [], can_interrupt: bool = false) -> void:
 					if can_interrupt:
 						escoria.event_manager.interrupt()
-
+					
```
Remove whitespace

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc_background.gd:
```diff
@@ -69,11 +69,12 @@ func _enter_tree():
 func _ready():
 	mouse_filter = MOUSE_FILTER_IGNORE
-
+	
```
Remove whitespace

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc_item.gd:
```diff
@@ -188,11 +188,13 @@ func _ready():
 	if not self.is_connected("mouse_exited", self, "_on_mouse_exited"):
 		connect("mouse_exited", self, "_on_mouse_exited")
-
+	
```
Remove whitespace

BHSDuncan commented on addons/escoria-core/game/core-scripts/esc_room.gd:
```diff
@@ -65,7 +65,7 @@ func _ready():
 			) != self.filename:
 		is_run_directly = true
-
+	
```
Remove whitespace

BHSDuncan reviewed (changes requested):
Just some whitespace to clean up. Otherwise the code looks ok to me.

balloonpopper commented:
Alt tab issue is fixed. Great work @StraToN !
